core/templates/components/summary-tile/lesson-card.component.spec.ts

```diff
@@ -0,0 +1,341 @@
+  };
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [FormsModule, HttpClientTestingModule],
+      declarations: [LessonCardComponent, MockTranslatePipe],
+      providers: [UrlInterpolationService, AssetsBackendApiService],
+      schemas: [NO_ERRORS_SCHEMA],
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(LessonCardComponent);
+    component = fixture.componentInstance;
+
+    urlInterpolationService = TestBed.inject(UrlInterpolationService);
+    assetsBackendApiService = TestBed.inject(AssetsBackendApiService);
+  });
+
+  it('should set story to CollectionSummary and its non-url values to the respective fields', () => {
+    component.story = CollectionSummary.createFromBackendDict(sampleCollection);
+
+    fixture.detectChanges();
+
+    expect(component.desc).toEqual(sampleCollection.objective);
+    expect(component.imgColor).toEqual(sampleCollection.thumbnail_bg_color);
+    expect(component.title).toEqual(sampleCollection.title);
+
+    expect(component.progress).toEqual(0);
+    expect(component.lessonTopic).toEqual('Collections');
+  });
+
+  it('should set story to LearnerExplorationSummary and its non-url values to the respective fields', () => {
+    component.story =
+      LearnerExplorationSummary.createFromBackendDict(sampleExploration);
+
+    fixture.detectChanges();
+
+    expect(component.desc).toEqual(sampleExploration.objective);
+    expect(component.imgColor).toEqual(sampleExploration.thumbnail_bg_color);
+    expect(component.title).toEqual(sampleExploration.title);
+
+    expect(component.progress).toEqual(0);
+    expect(component.lessonTopic).toEqual('Community Lessons');
+  });
+
+  it('should set story to complete StorySummary and its non-url values to the respective fields', () => {
+    component.story = StorySummary.createFromBackendDict(sampleTopic);
+    component.topic = sampleTopic.topic_name;
+
+    fixture.detectChanges();
+
+    expect(component.desc).toEqual(sampleTopic.title);
+    expect(component.imgColor).toEqual(sampleTopic.thumbnail_bg_color);
+
+    let nextStory =
+      sampleTopic.completed_node_titles.length -
+      (sampleTopic.completed_node_titles.length ===
+      sampleTopic.node_titles.length
+        ? 1
+        : 0);
+
+    expect(component.title).toEqual(
+      `Chapter ${nextStory + 1}: ${sampleTopic.node_titles[nextStory]}`
+    );
+    expect(component.progress).toEqual(
+      Math.floor(
+        (sampleTopic.completed_node_titles.length /
+          sampleTopic.node_titles.length) *
+          100
+      )
+    );
+    expect(component.lessonTopic).toEqual(sampleTopic.topic_name);
+  });
+
+  it('should set story to incomplete StorySummary and its non-url values to the respective fields', () => {
+    component.story = StorySummary.createFromBackendDict(incompleteTopic);
+    component.topic = incompleteTopic.topic_name;
+
+    fixture.detectChanges();
+
+    let nextStory =
```

Comment:
1. let->const
2. This isn't a good way to write a test. You are putting logic that you try to test in the test itself. Such tests usually pass, but they don't test the logic, they test your ability to replicate code.

Use concrete numbers here, you have all the data to calculate nextStory and set it to a specific number, don't you?

Same goes for the test above and any other test that involves logic. Try to write tests without logic. Test should verify that provided some inputs a module (function/class/...) returns specific outputs, where outputs should be written explicitly, not calculated based on the inputs.

The problem with putting logic into tests is that this way there's a high chance you transfer your bug from logic into tests, tests will pass despite a bug in the implementation.

Reply:
Replaced with concrete values. Done.